text/0000-unstable-sort.md

```diff
@@ -0,0 +1,213 @@
+- Feature Name: unstable_sort
+- Start Date: 2017-02-03
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Add an unstable sort to libcore.
+
+# Motivation
+[motivation]: #motivation
+
+At the moment, the only sort function we have in libstd is `slice::sort`. It is stable,
+allocates additional memory, and is unavailable in `#![no_std]` environments.
+
+**Q: What is stability?**<br>
+A: A sort function is stable if it doesn't reorder equal elements. For example:
+```rust
+let mut orig = vec![(0, 5), (0, 4)];
+let mut v = orig.clone();
+
+// Stable sort preserves the original order of equal elements.
+v.sort_by_key(|p| p.0);
+assert!(orig == v); // OK!
+
+/// Unstable sort may or may not preserve the original order.
+v.unstable_sort_by_key(|p| p.0);
+assert!(orig == v); // MAY FAIL!
+```
+
+**Q: When is stability useful?**<br>
+A: Not very often. A typical example is sorting columns in interactive GUI tables.
+If you want to have rows sorted by column X while breaking ties by column Y, you
+first have to click on column Y and then click on column X.
+
+**Q: Can stable sort be performed using unstable sort?**<br>
+A: Yes. If we transform `[T]` into `[(T, usize)]` by pairing every element with it's
+index, then perform unstable sort, and finally remove indices, the result will be
+equivalent to stable sort.
+
+**Q: Why is `slice::sort` stable?**<br>
+A: Because stability is a good default. A programmer might call a sort function
+without checking in the documentation whether it is stable or unstable. It is very
+inutitive to assume stability, so having `slice::sort` perform unstable sorting might
+cause unpleasant surprises.
+
+**Q: Why does `slice::sort` allocate?**<br>
+A: It is possible to implement a non-allocating stable sort, but it would be
+considerably slower.
+
+**Q: Why is `slice::sort` not compatible with `#![no_std]`?**<br>
+A: Because it allocates additional memory.
+
+**Q: How much faster can unstable sort be?**<br>
+A: Sorting 64-bit integers using [pdqsort][stjepang-pdqsort] (an
+unstable sort implementation) is **40% faster** than using `slice::sort`.
+Detailed benchmarks are [here](https://github.com/stjepang/pdqsort#extensive-benchmarks).
```

Comment:
Out of curiosity, is this compared against the recent improvements to the existing sort implementation?